File: JAVA/1074. Number of Submatrices That Sum to Target.java
class Solution {
    public int numSubmatrixSumTarget(int[][] matrix, int target) {
        final int m = matrix.length;
    final int n = matrix[0].length;
    int res = 0;
    for (int[] row : matrix)
      for (int i = 1; i < n; ++i)
        row[i] += row[i - 1];
    for (int baseCol = 0; baseCol < n; ++baseCol)
      for (int j = baseCol; j < n; ++j) {
        Map<Integer, Integer> prefixCount = new HashMap<>();
        prefixCount.put(0, 1);
        int sum = 0;
        for (int i = 0; i < m; ++i) {
          if (baseCol > 0)
            sum -= matrix[i][baseCol - 1];
          sum += matrix[i][j];
          res += prefixCount.getOrDefault(sum - target, 0);
          prefixCount.merge(sum, 1, Integer::sum);
        }
      }
    return res;
    }
}

File: C++/1074. Number of Submatrices That Sum to Target.cpp
class Solution {
public:
int subarraySum(vector<int>& nums, int k) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);        
        int num = 0, curr_sum = 0;
        unordered_map<int, int>mp;
        for(int i=0; i<nums.size(); i++)
        {
            curr_sum += nums[i];
            if(curr_sum == k)num++;
            if(mp.find(curr_sum - k) != mp.end())
            {
                num += mp[curr_sum - k];
            }
            mp[curr_sum]++;
        }
        return num;
    }    
    int numSubmatrixSumTarget(vector<vector<int>>& mat, int target) {
        int total=0;
        int n = mat.size();
        int m = mat[0].size();
        for(int i=0; i<n; i++){			
            vector<int> sum(m, 0);
            for(int j=i; j<n; j++){
                for(int k=0; k<m; k++){
                    sum[k] += mat[j][k];
                }
                total += subarraySum(sum, target);
            }
        }       
    return total;
        
    }
};
